### PR TITLE
feat(api-client): improve oauth2 error handling + add PKCE capabilities with tests

### LINKED ISSUE #604: Oauth 2.0 is unfinished, PKCE is missing
### Expected Behavior

oauth2.0 to be functional

### Actual Behavior

image

### Reproduction steps

```
on latest main with oauth2.0 as an auth type, it's missing and has warning
```

### Link to site / OpenAPI Spec

[docs.scalar.com/swagger-editor](http://docs.scalar.com/swagger-editor)

### Version

latest

### What browsers are you seeing the problem on?

*No response*

<sub>From [SyncLinear.com](https://synclinear.com) | [DOC-1382](https://linear.app/api-docs/issue/DOC-1382/oauth-20-is-unfinished)</sub>

### LINKED ISSUE #3705: support x-usePkce
Add support for x-usePkce https://redocly.com/docs/api-reference-docs/specification-extensions/x-use-pkce

### PR DESCRIPTION
**Pull request by amritk**
closes #604
closes #3705

Added a super basic front-end
image

and tested it using spotify (they support pkce)
image

### CODE REVIEW

File: packages/api-client/src/views/Request/libs/oauth2.ts [hunk 1]
```diff
@@ -16,45 +17,108 @@ export const isOauth2Example = (
...
  */
-export const authorizeOauth2 = (
+export const authorizeOauth2 = async (
   scheme: SecuritySchemeOauth2,
   example: SecuritySchemeOauth2ExampleValue,
   /** We use the active server to set a base for relative redirect uris */
   activeServer: Server,
-) =>
-  new Promise<string>((resolve, reject) => {
+): Promise<ErrorResponse<string>> => {
```
Comment (amritk) [inline packages/api-client/src/views/Request/libs/oauth2.ts:68]: I had to change this around anyway to handle some await crypto action so I ended up switching it to the ErrorResponse style for some better error handling

File: packages/api-client/src/views/Request/libs/oauth2.test.ts [hunk 2]
```diff
@@ -147,14 +150,98 @@ describe('oauth2', () => {
...
+          for (let i = 0; i < arr.length; i++) {
+            arr[i] = i
+          }
+        }
+        return arr
+      })
+
+      // Mock crypto.subtle.digest
+      vi.spyOn(crypto.subtle, 'digest').mockResolvedValue(
+        new Uint8Array([1, 2, 3, 4, 5, 6, 8, 9, 10]).buffer,
```
Comment (big-kahuna-burger) [inline packages/api-client/src/views/Request/libs/oauth2.test.ts:178]: This should be 32 bytes.

File: packages/api-client/src/views/Request/libs/oauth2.ts [hunk 3]
```diff
@@ -16,45 +17,108 @@ export const isOauth2Example = (
 ): example is SecuritySchemeOauth2ExampleValue =>
   example.type.startsWith('oauth')
 
+type PKCEState = {
+  codeVerifier: string
+  codeChallenge: string
+  codeChallengeMethod: string
```
Comment (big-kahuna-burger) [inline packages/api-client/src/views/Request/libs/oauth2.ts:23]: According to proto there is 'plain' and 'S256'. Make 'S256' default if not explicitly provided  plain. Plain is discouraged but some providers don't support S256, so make it option.

**Comment by big-kahuna-burger** [comment]
https://github.com/panva/oauth4webapi/blob/main/src/index.ts#L1386C1-L1388C2

See more here.

File: packages/api-client/src/views/Request/libs/oauth2.test.ts (hunk 2 above)
Comment (amritk) [reply packages/api-client/src/views/Request/libs/oauth2.test.ts:178]: Would it matter if its just a mock? It really doesn't do much

Comment (amritk) [reply packages/api-client/src/views/Request/libs/oauth2.test.ts:178]: Nvm got it will change the generator

**Comment by amritk** [comment]
@hanspagel maybe you can ping me on what didn't work and we can get it fixed!

**Pull request merged**